src/ray/gcs/subscribe_executor.cc

```diff
@@ -0,0 +1,106 @@
+#include "ray/gcs/subscribe_executor.h"
+
+namespace ray {
+
+namespace gcs {
+
+template <typename ID, typename Data, typename Table>
+Status SubscribeExecutor<ID, Data, Table>::AsyncSubscribe(
+    const ClientID &client_id, const SubscribeCallback<ID, Data> &subscribe,
+    const StatusCallback &done) {
+  // TODO(micafan) Optimize the lock when necessary.
+  std::lock_guard<std::mutex> lock(mutex_);
+
+  if (registered_) {
+    return Status::OK();
+  }
+
+  auto on_subscribe = [this, subscribe](RedisGcsClient *client, const ID &id,
+                                        const std::vector<Data> &result) {
+    if (result.empty()) {
+      return;
+    }
+
+    SubscribeCallback<ID, Data> callback = nullptr;
+    {
+      std::lock_guard<std::mutex> lock(mutex_);
+      const auto it = id_to_request_map_.find(id);
+      if (it != id_to_request_map_.end()) {
+        callback = it->second.subscribe_;
+      }
+    }
+    if (callback != nullptr) {
+      callback(id, result.back());
+    } else if (subscribe != nullptr) {
+      subscribe(id, result.back());
+    }
+  };
+
+  auto on_done = [this, done](RedisGcsClient *client) {
+    if (done != nullptr) {
+      done(Status::OK());
+    }
+  };
+
+  Status status = table_.Subscribe(JobID::Nil(), client_id, on_subscribe, on_done);
+  if (status.ok()) {
+    registered_ = true;
+  }
+
+  return status;
+}
+
+template <typename ID, typename Data, typename Table>
+Status SubscribeExecutor<ID, Data, Table>::AsyncSubscribe(
+    const ClientID &client_id, const ID &id, const SubscribeCallback<ID, Data> &subscribe,
+    const StatusCallback &done) {
+  Status status = AsyncSubscribe(client_id, nullptr, nullptr);
+  if (!status.ok()) {
+    return status;
```

Comment:
this doesn't look right... in this case subscribing the second actor would fail, because the call to `AsyncSubscribe` would return failure as it has been registered

Reply:
and we should probably add a check that client_id is not NIL, or return failure in that case?

Reply:
> Some scenarios require a valid ClientID to be passed in. e.g.: TaskLeaseTable and TaskTable.

Don't quite understand here. I think when subscribe for a specific actor, then it always needs a valid `client_id`, otherwise it won't work correctly.

Reply:
In `ActorTable` you can always read/subscribe/write an `actor` without `ClientID`.  So we can remove `ClientID` from `ActorStateAccessor`'s API. And pass `Nil` as `ClientID` to `ActorTable`.

The reason why `ActorTable`'s  API still has `ClientID` is `ActorTable`'s parent class `Log` need it. Some derived class of `Log` (such as `TaskTable`) need a valid `ClientID`.  An historical omission here.

Reply:
Thanks. My original comment just we need to use the client_id when subscribing for a specific actor, as in that case we want redis to publish updates of the specific key to this client, this can be either passed via an arg, or embedded inside the gcs client though.